core/controller/src/main/scala/whisk/core/controller/actions/PrimitiveActions.scala

```diff
@@ -139,6 +199,356 @@ protected[actions] trait PrimitiveActions {
+          completeAppActivation(user, session, response)
+        case Success(Left(activationId)) =>
+          // invocation timeout
+          session.logs += activationId
+          val response = ActivationResponse.whiskError(compositionActivationTimeout(activationId))
+          completeAppActivation(user, session, response)
+        case Success(Right(activation)) =>
+          // successful invocation
+          session.logs += activation.activationId
+          session.duration += activation.duration.getOrElse(activation.end.toEpochMilli - activation.start.toEpochMilli)
+
+          val result = activation.resultAsJson
+
+          // extract params from result
+          val params = result.getFields(WhiskActivation.paramsField).headOption.map { p =>
+            Try(p.asJsObject).getOrElse(JsObject(WhiskActivation.valueField -> p)) // ensure params is a dictionary
+          }
+
+          // update session state
+          session.state = result.getFields(WhiskActivation.stateField).headOption.flatMap { p =>
+            Try(Some(p.asJsObject)).getOrElse(None)
+          }
+
+          // extract next action from result and invoke
+          result.getFields(WhiskActivation.actionField).headOption match {
+            case None =>
+              // no next action, end composition execution, return to caller
+              val response = ActivationResponse(activation.response.statusCode, Some(params.getOrElse(result)))
+              completeAppActivation(user, session, response)
+            case Some(next) =>
+              Try(next.convertTo[EntityPath]) match {
+                case Failure(t) =>
+                  // parsing failure
+                  val response = ActivationResponse.applicationError(compositionComponentInvalid(next))
+                  completeAppActivation(user, session, response)
+                case Success(next) =>
+                  // resolve and invoke next action
+                  val fqn = (if (next.defaultPackage) EntityPath.DEFAULT.addPath(next) else next)
+                    .resolveNamespace(user.namespace)
+                    .toFullyQualifiedEntityName
+                  val resource = Resource(fqn.path, Collection(Collection.ACTIONS), Some(fqn.name.asString))
+                  entitlementProvider.check(user, Privilege.ACTIVATE, Set(resource), noThrottle = true).onComplete {
+                    case Failure(t) =>
+                      // failed entitlement check
+                      val response =
+                        ActivationResponse.applicationError(compositionComponentNotAccessible(next.toString))
+                      completeAppActivation(user, session, response)
+                    case Success(_) =>
+                      // successful entitlement check
+                      WhiskActionMetaData.resolveActionAndMergeParameters(entityStore, fqn).onComplete {
+                        case Failure(t) =>
```

Comment:
Here, the resolve and merge operation if it fails should be an instance of an ArtifactStore exception, which may be a user error (action doesn't exist), or possibly an internal error (db connection partition). So some of these errors should be disambiguated for whisk errors.